### PR TITLE
add feature_engineering from robynpy_interfaces

### PR DESCRIPTION
**Pull request by alxlyj**
## Project Robyn

**Summary:** We are including the interfaces (methods and data classes) for the modeling subcomponent in Python. This interface is inferred from inputs.R.  

## What is new:
- feat: New Python interfaces for modeling subcomponent. 

### Class diagram
```mermaid
classDiagram
    class FeatureEngineering {
        +MMMData mmm_data
        +Hyperparameters hyperparameters
        +perform_feature_engineering(quiet: bool) FeaturizedMMMData
        -_prepare_data() DataFrame
        -_create_rolling_window_data(dt_transform: DataFrame) DataFrame
        -_calculate_media_cost_factor(dt_input_roll_wind: DataFrame) Series
        -_run_models(dt_input_roll_wind: DataFrame, media_cost_factor: Series) Dict
        -_fit_spend_exposure(dt_spend_mod_input: DataFrame, media_cost_factor: float, paid_media_var: str) Dict
        -_prophet_decomposition(dt_transform: DataFrame) DataFrame
        -_set_holidays() DataFrame
        -_check_no_variance(dt_transform: DataFrame) None
        +hyper_names(adstock: str, all_media: List[str], all_vars: List[str])* List[str]
        +hyper_limits()* DataFrame
    }
    
    class FeaturizedMMMData {
        +DataFrame dt_mod
        +DataFrame dt_modRollWind
        +Dict modNLS
    }
    
    class FeaturePlotter {
        +MMMData mmm_data
        +Hyperparameters hyperparameters
        +plot_adstock(channel: str) Figure
        +plot_saturation(channel: str) Figure
        +plot_spend_exposure(dt_mod: DataFrame, channel: str) Figure
        +plot_feature_importance(feature_importance: Dict[str, float]) Figure
        +plot_response_curves(dt_mod: DataFrame) Dict[str, Figure]
        +save_plots(plots: Dict[str, Figure], output_dir: str) None
        -_apply_transformations(x: ndarray, params: ChannelHyperparameters) ndarray
    }
    
    class MMMData {
        +DataFrame data
        +MMMDataSpec mmmdata_spec
    }
    
    class MMMDataSpec {
        +str dep_var
        +DependentVarType dep_var_type
        +str date_var
        +List[str] paid_media_spends
        +List[str] paid_media_vars
        +List[str] organic_vars
        +List[str] context_vars
        +List[str] prophet_vars
    }
    
    class Hyperparameters {
        +Dict[str, ChannelHyperparameters] hyperparameters
        +AdstockType adstock
        +float lambda_
        +List[float] train_size
        +get_hyperparameter(channel: str) ChannelHyperparameters
        +has_channel(channel: str) bool
        +get_hyperparameter_limits()* Dict[str, List[str]]
    }
    
    class ChannelHyperparameters {
        +List[float] thetas
        +List[float] shapes
        +List[float] scales
        +List[float] alphas
        +List[float] gammas
        +List[bool] penalty
    }
    
    FeatureEngineering ..> FeaturizedMMMData : produces
    FeatureEngineering --> MMMData : uses
    FeatureEngineering --> Hyperparameters : uses
    FeaturePlotter --> MMMData : uses
    FeaturePlotter --> Hyperparameters : uses
    MMMData --> MMMDataSpec : contains
    Hyperparameters --> ChannelHyperparameters : contains
```

### Parity:
```mermaid
graph TB
    subgraph R["R Implementation (inputs.R)"]
        R1[robyn_inputs]
        R2[robyn_engineering]
        R3[prophet_decomp]
        R4[fit_spend_exposure]
        R5[set_holidays]
        R6[hyper_names]
        R7[hyper_limits]
        R8[plot_adstock]
        R9[plot_saturation]
    end

    subgraph Python["Python Implementation"]
        subgraph FE["feature_engineering.py"]
            P1[FeatureEngineering class]
            P2[perform_feature_engineering]
            P3[_prepare_data]
            P4[_create_rolling_window_data]
            P5[_calculate_media_cost_factor]
            P6[_run_models]
            P7[_fit_spend_exposure]
            P8[_prophet_decomposition]
            P9[_set_holidays]
            P10[_check_no_variance]
            P13[hyper_names]
            P14[hyper_limits]
            P15[FeaturizedMMMData]
        end
        
        subgraph PL["plotter.py"]
            Q1[FeaturePlotter class]
            Q2[plot_adstock]
            Q3[plot_saturation]
            Q4[plot_spend_exposure]
            Q5[plot_feature_importance]
            Q6[plot_response_curves]
        end
    end

    R1 -->|Calls| R2
    R2 -->|Uses| R3
    R2 -->|Uses| R4
    R2 -->|Uses| R5
    R1 -.->|Uses| R6
    R1 -.->|Uses| R7
    R1 -.->|Can use| R8
    R1 -.->|Can use| R9

    P1 -->|Contains| P2
    P2 -->|Calls| P3
    P2 -->|Calls| P4
    P2 -->|Calls| P5
    P2 -->|Calls| P6
    P6 -->|Uses| P7
    P2 -->|Calls| P8
    P8 -->|Uses| P9
    P2 -->|Calls| P10
    P1 -->|Contains| P13
    P1 -->|Contains| P14
    P2 -->|Returns| P15

    Q1 -->|Contains| Q2
    Q1 -->|Contains| Q3
    Q1 -->|Contains| Q4
    Q1 -->|Contains| Q5
    Q1 -->|Contains| Q6

    R1 -.->|Ported to| P1
    R2 -.->|Ported to| P2
    R3 -.->|Ported to| P8
    R4 -.->|Ported to| P7
    R5 -.->|Ported to| P9
    R6 -.->|Ported to| P13
    R7 -.->|Ported to| P14
    R8 -.->|Ported to| Q2
    R9 -.->|Ported to| Q3

    classDef rClass fill:#f9d5e5,stroke:#333,stroke-width:2px;
    classDef pyClass fill:#eeeeee,stroke:#333,stroke-width:2px;
    classDef plotterClass fill:#e6f3ff,stroke:#333,stroke-width:2px;
    class R1,R2,R3,R4,R5,R6,R7,R8,R9 rClass;
    class P1,P2,P3,P4,P5,P6,P7,P8,P9,P10,P13,P14,P15 pyClass;
    class Q1,Q2,Q3,Q4,Q5,Q6 plotterClass;


```

-  No implementations have been included just yet with this PR so it would be lightweight to land. More rigorous testing would be included in the implementation PR.

### CODE REVIEW

File: python/src/robyn/modeling/feature_engineering.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            all_media (List[str]): List of all media variables.
+            all_vars (Optional[List[str]], optional): List of all variables. Defaults to None.
+
+        Returns:
+            List[str]: The names of all hyperparameters.
+        """
+        pass
+
+    @staticmethod
+    def hyper_limits() -> pd.DataFrame:
```
Comment (sumane81) [inline python/src/robyn/modeling/feature_engineering.py:200]: data.entities.Hyperparameters may be better place for this.

File: python/src/robyn/modeling/feature_engineering.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        data preparation, rolling window creation, media cost factor calculation,
+        model running, and prophet decomposition (if applicable).
+
+        Args:
+            quiet (bool, optional): If True, suppresses print statements. Defaults to False.
+
+        Returns:
+            FeaturizedMMMData: The output of the feature engineering process.
+        """
+        if not quiet:
```
Comment (sumane81) [inline python/src/robyn/modeling/feature_engineering.py:77]: No quiet flags or no print statements anywhere. We will be using logging framework that is already added with file handler (prints into log file ) and console handler (prints onto console). Console handler can be updated to remove datetime prefix if we want it cleaner.

**Review by sumane81 — CHANGES_REQUESTED** [review]
1. We started using logging framework so no more print statements
2. Move FeaturePlotter to visualization
3. I see some implementation here, which is okay but subsequent PR needs to be added with unit tests. I assume you will add unit tests for all implementation in next PR.

File: python/src/robyn/modeling/feature_engineering.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    def _set_holidays(self) -> pd.DataFrame:
+        """
+        Set up the holidays for the Prophet model.
+
+        Returns:
+            pd.DataFrame: The DataFrame with holidays.
+        """
+        pass
+
+    def _check_no_variance(self, dt_transform: pd.DataFrame) -> None:
```
Comment (sumane81) [inline python/src/robyn/modeling/feature_engineering.py:175]: No validations happens here.

File: python/src/robyn/modeling/feature_engineering.py (hunk 2 above)
Comment (alxlyj) [reply python/src/robyn/modeling/feature_engineering.py]: removed - will move integration with logging framework as part of implementation PR.

File: python/src/robyn/modeling/feature_engineering.py (hunk 1 above)
Comment (alxlyj) [reply python/src/robyn/modeling/feature_engineering.py]: removed as these are already available as static methods within HyperParameter #1043

**Review by alxlyj — COMMENTED** [review]
addressed comments.

File: python/src/robyn/visualization/plotter.py
Comment (sumane81) [inline python/src/robyn/visualization/plotter.py:1]: we will have multiple plotter classes. can you be specific in file name based on kinds of plots it creates?

File: python/src/robyn/visualization/plotter.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            featurized_data (FeaturizedMMMData): The featurized data after feature engineering.
+
+        Returns:
+            Dict[str, plt.Figure]: Dictionary mapping channel names to their respective response curve plots.
+        """
+        dt_mod = featurized_data.dt_mod
+        # Rest of the method implementation
+        pass
+
+    def save_plots(self, plots: Dict[str, plt.Figure], output_dir: str) -> None:
```
Comment (sumane81) [inline python/src/robyn/visualization/plotter.py:91]: can you remove this for now? create_all_plots that will create all plots and can take argument to export is better than this method, which takes plots that may or may not be created by this class and just exports it. But let's not even add it for now. robyn/utils/data_persistence_util.py will be single entry point for reading and writing to persistence layer.

**Comment by shivkanthb** [comment]
Overall looks pretty good to me

File: python/src/robyn/visualization/plotter.py
Comment (alxlyj) [reply python/src/robyn/visualization/plotter.py:1]: renamed to feature_plotter.py

File: python/src/robyn/visualization/plotter.py (hunk 4 above)
Comment (alxlyj) [reply python/src/robyn/visualization/plotter.py:91]: removing save_plots - to be used for data_persistence_util.py when we work on this interface.

**Review by sumane81 — APPROVED** [review]

**Pull request merged**